### PR TITLE
remove math kernels that have derivative formulas in core

### PR DESCRIPTION
**Pull request by bdhirsh**
Some ops already have a derivative formula in core, but have registered `CompositeImplicitAutograd` kernels. This PR makes them `CompositeExplicitAutograd` instead, which means:
- backends will now use the derivative formulas in core, instead of running autograd on their decompositions
- transforms (vmap / functionalize) will treat these ops as "primitives" instead of decomposing them

The second point fixes a numeric accuracy bug in the mobile pipeline: You end up with numeric differences when you run the following:
```
torch.ops.layer_norm.default
functionalize(torch.ops.layer_norm.default)
```

Because the second will use the decomposition of layer_norm, while the first won't.

Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78183

### CODE REVIEW

**Review by ezyang — APPROVED** [review]

**Comment by Chillee** [comment]
Is there any reason we don't just remove them?

**Comment by bdhirsh** [comment]
Rip, looks like XLA tests are failing though, because XLA doesn't have a lowering for `native_layer_norm_backward`.

I wonder if we can write a composite kernel for `native_layer_norm_backward` in terms of `native_batch_norm_backward`...

**Comment by ezyang** [comment]
We have it already, just need xla to use it

**Comment by bdhirsh** [comment]
> We have it already, just need xla to use it

We do?

We have one for `native_layer_norm`, but I don't see one for its backward: https://github.com/pytorch/pytorch/blob/a95f1edd8549b6a249ffa448df073ac4c8b81382/aten/src/ATen/native/native_functions.yaml#L2685

I should be able to figure out how to write a composite one though, just by looking at the composite `native_layer_norm` kernel and manually writing out the backward of each step?

**Comment by ezyang** [comment]
It's written in Python in torch/_decomps. We can just register it to XLA key too

**Comment by bdhirsh** [comment]
> It's written in Python in torch/_decomps. We can just register it to XLA key too

Awesome :) I was painstakingly trying to figure out the backward decomp from the forward one in C++

**Comment by bdhirsh** [comment]
At first I was going to try to arrange for us to register the decomp to the `CompositeExplicitAutograd` key at `import torch` time. That doesn't actually work though, because we get a circular dependency (all of the existing python decomp init code runs after `import torch`).

I guess instead we can update `pytorch/xla` to directly register the decomp to it's key when you `import pytorch_xla` (and risk that other non-XLA backend weren't relying on it).

**Comment by bdhirsh** [comment]
We also need one for `native_group_norm_backward` (looks like XLA will need it, but we don't have a python decomp for it.

Maybe a good time to add one :p

**Comment by ezyang** [comment]
Cc @Chillee

**Comment by Chillee** [comment]
@bdhirsh Should we use the Torchscript-based way of calling decomps for now? I'm a bit worried the `__torch_dispatch__` overhead will be prohibitive for XLA.

**Comment by bdhirsh** [comment]
> @bdhirsh Should we use the Torchscript-based way of calling decomps for now? I'm a bit worried the `__torch_dispatch__` overhead will be prohibitive for XLA.

That's a good point, we should probably use the torchscript path. Do you have a code pointer to where it's used in functorch somewhere?

**Comment by Chillee** [comment]
JIT boxed fallback: https://github.com/pytorch/functorch/blob/main/functorch/csrc/BatchRulesHelper.cpp#L136

python registration: https://github.com/pytorch/functorch/blob/main/functorch/_src/eager_transforms.py#L1278

vmap binding: https://github.com/pytorch/functorch/blob/main/functorch/csrc/BatchRulesViews.cpp#L510

cc: @eellison

**Comment by bdhirsh** [comment]
ok, semi-yolo xla side PR is here: https://github.com/pytorch/xla/pull/3612. I'll check the CI for it.

Separately, @ZolotukhinM mentioned that fixing the `native_layer_norm` issue is pretty important for mobile, but `native_group_norm` isn't an active blocker. Given that we don't have a decomposition for `native_group_norm_backward`, I'll just land the fix for layer_norm for now (once I see green XLA CI).
